pkg/analyzer/lib/src/generated/gn.dart

```diff
@@ -243,26 +155,56 @@ class GnWorkspace extends Workspace {
     return null;
   }
 
-  /**
-   * Locate the Dart sources directory.
-   *
-   * Return `null` if it could not be found.
-   */
-  static String _getPackagesDirectory(ResourceProvider provider, String root) {
+  static String _findPackagesFile(
+      ResourceProvider provider, String root, String path,
+      {forHost: false}) {
     Context pathContext = provider.pathContext;
-    String outDirectory = provider
-        .getFolder(pathContext.join(root, 'out'))
+    String sourceDirectory = pathContext.relative(path, from: root);
+    Folder outDirectory = provider.getFolder(pathContext.join(root, 'out'));
+    if (!outDirectory.exists) {
+      return null;
+    }
+    outDirectory = outDirectory
         .getChildren()
         .where((resource) => resource is Folder)
         .map((resource) => resource as Folder)
         .firstWhere((Folder folder) {
-      String baseName = basename(folder.path);
+      String baseName = pathContext.basename(folder.path);
       // TODO(pylaligand): find a better way to locate the proper directory.
       return baseName.startsWith('debug') || baseName.startsWith('release');
-    }, orElse: () => null)?.path;
-    return outDirectory == null
-        ? null
-        : provider.pathContext.join(outDirectory, 'gen', 'dart.sources');
+    }, orElse: () => null);
```

Comment:
This particular sequence of invocations appears in three places. Given that we control the API of Folder, we should add some utility methods that do some of this work rather than repeating it. (Not a request for a change to this CL, though.)